### PR TITLE
Fix wrong behavior on s390x

### LINKED ISSUE #86: Wrong behavior on s390x
We are currently facing an issue using this library on s390x. I tired to run the test `TestProgramRun`. I got:
```sh
go test -run TestProgramRun
--- FAIL: TestProgramRun (0.00s)
    prog_test.go:50:     0: LdXMemW dst: r2 src: r1 off: 4 imm: 0
             1: LdXMemW dst: r1 src: r1 off: 0 imm: 0
             2: MovReg dst: r3 src: r1
             3: AddImm dst: r3 imm: 14
             4: JGTReg dst: r3 off: -1 src: r2 <out>
             5: StMemB dst: r1 src: r0 off: 0 imm: 222
             6: StMemB dst: r1 src: r0 off: 1 imm: 173
             7: StMemB dst: r1 src: r0 off: 2 imm: 190
             8: StMemB dst: r1 src: r0 off: 3 imm: 239
        out:
             9: LdImmDW dst: r0 imm: 42
            11: Exit
        
    prog_test.go:59: can't load program: permission denied: 0: (61) r1 = *(u32 *)(r2 +4)
        R2 !read_ok
        processed 1 insns (limit 1000000) max_states_per_insn 0 total_states 0 peak_states 0 mark_read 0
FAIL
exit status 1
FAIL    github.com/cilium/ebpf  0.001s
Collapse
```

### PR DESCRIPTION
**Pull request by iii-i**
Fixes #86

### CODE REVIEW

File: internal/testutils/endianness.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package testutils
+
+import (
+	"unsafe"
+)
+
+func GetHostEndianness() string {
```
[the remark this answers is not in the document]
Comment (iii-i) [reply internal/testutils/endianness.go:7]: Will do.

File: testdata/Makefile [hunk 2]
```diff
@@ -1,19 +1,22 @@
...
 
 clean:
 	-$(RM) *.elf
 
-loader-%.elf: loader.c
-	$* -target bpf -O2 -g \
-		-Wall -Werror \
-		-c $< -o $@
+loader-%-el.elf: loader.c
+	$* $(CFLAGS) -mlittle-endian -c $< -o $@
```
[the remark this answers is not in the document]
Comment (iii-i) [reply testdata/Makefile:13]: Not sure, but kernel selftests use `-mlittle-endian`/`-mbig-endian`.

File: asm/instruction_systemz.go [hunk 3]
```diff
@@ -0,0 +1,15 @@
+// +build s390 s390x
```
[the remark this answers is not in the document]
Comment (iii-i) [reply asm/instruction_systemz.go:1]: Hmm, in theory s390 should be there: https://github.com/golang/go/blob/master/src/go/build/syslist.go
Maybe distro builds somehow disable it?

File: asm/instruction_systemz.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+package asm
+
+func newBPFRegisters(dst, src Register) bpfRegisters {
```
[the remark this answers is not in the document]
Comment (iii-i) [reply asm/instruction_systemz.go:5]: Would it be ok if I implement an s390 specific test for this? The reason is that I'm not sure how other big endian arches handle bitfields. On the one hand side it should be reasonable to expect that it should be all the same everywhere, but technically it must be arch-specific ABI document that mandates bit field order.

[the remark this answers is not in the document]
Comment (iii-i) [reply asm/instruction_systemz.go:5]: OTOH LLVM is using endianness information to determine bit field order: https://github.com/llvm/llvm-project/blob/llvmorg-10.0.0/llvm/lib/Target/BPF/MCTargetDesc/BPFMCCodeEmitter.cpp#L113

So if there is some weird big-endian arch that a different bit field order, we are in trouble anyway. I'll replace all s390 checks with endianness checks.

**Comment by lmb** [comment]
Your proposals sound sane, looking forward to your update!

**Comment by iii-i** [comment]
I actually pushed one a few days back (3db7828) :-) Could you have a look please?

**Comment by lmb** [comment]
Oh boy, sorry about that! I missed the notification somehow. Thanks for working on this!

**Pull request merged**